### DISCUSSION TITLE
Question about `chezmoi update` and SSH config

### DISCUSSION THREAD

**Discussion by xfzv** [twpayne/chezmoi#3185] [Q&A]
I have set up SSH keys on my different machines lately.

On my desktop, where my `chezmoi` source directory is located, I have two Linux distros installed. The desktop has a static local IP address.

To easily ssh into the desktop from my other machines, I'm using the following in `~/.ssh/config` (here, the one from a VM):

```
Host archlinux-desktop
  HostName <desktop_ip_address>
  HostKeyAlias archlinux-desktop
  IdentityFile ~/.ssh/archlinux-vm

Host gentoo-desktop
  HostName <desktop_ip_address>
  HostKeyAlias gentoo-desktop
  IdentityFile ~/.ssh/archlinux-vm
```

the above allows me to use either:

```console
❯ ssh user@archlinux-desktop
❯ ssh user@gentoo-desktop
```

from the VM, depending on which distro is running on my desktop. It works fine.

However, when I had initialized the `chezmoi` source directory on the VM, the SSH keys weren't set up so I used:

```bash
❯ chezmoi --source "~/dotfiles" init --ssh user@desktop_ip_address:/path/to/my/bare/repository
```

As a result, now when I run `chezmoi update` from the VM, I guess `chezmoi` is still looking for `<desktop_ip_address>` in `~/.ssh/known_hosts` instead of either `archlinux-desktop` or `gentoo-desktop`:

```console
❯ chezmoi update
The authenticity of host '<desktop_ip_address> (<desktop_ip_address>)' can't be established.
ED25519 key fingerprint is SHA256:redacted.
This host key is known by the following other names/addresses:
    ~/.ssh/known_hosts:4: gentoo-desktop
Are you sure you want to continue connecting (yes/no/[fingerprint])?
```

How should I configure `chezmoi` on the VM so that it can pull changes from the bare repository on my desktop, independently of which distro is booted on the desktop?

I thought I could just do:

```bash
# Delete the source directory on the VM
❯ rm -rf ~/dotfiles

# Run chezmoi init again with the host name instead of the IP address
❯ chezmoi --source "~/dotfiles" init --ssh user@<currently_booted_distro>-desktop:/path/to/my/bare/repository
```

but I assume that if I use `gentoo-desktop` in the above command, then I won't be able to run `chezmoi update` if my desktop is booted with Arch Linux and vice versa?

**Comment by halostatue** [comment]
Check the output of `git remote -v` in your chezmoi source directory. Chezmoi is doing a `git clone user@host:/path/to/my/bare/repository` and that never gets updated.

I’m not entirely sure that what you want is going to be able to happen, because it’s a *git* issue, not really a *chezmoi* issue.

**Reply by xfzv** [reply]
```console
❯ git remote -v
origin user@<desktop_ip_address>:/path/to/my/bare/repository (fetch)
origin user@<desktop_ip_address>:/path/to/my/bare/repository (push)
```

> it’s a _git_ issue, not really a _chezmoi_ issue.

Indeed. I did a quick research about adding remotes to a git repository. It might work when pulling/pushing with `git` commands by manually mentioning the remote name, but I'm not sure if it would work with `chezmoi update` command. Plus, I wonder what would happen considering one of the two remotes would never be reachable (only one OS booted on my desktop).

What I ended up doing for now is to keep the current setup with `user@<desktop_ip_address>`. To circumvent the `"Warning: Remote Host Identification Has Changed”` error when running `chezmoi update` from the VM if my desktop currently booted OS has changed since last `chezmoi update`, I added the following to `~/.ssh/config`:

```
Host <desktop_ip_address>
  IdentityFile ~/.ssh/archlinux-vm
```

And I edited my original `chu` alias:

```bash
if rg -q "vm" "/etc/hostname"; then
  chu(){
    if rg -q "desktop_ip_address" "$HOME/.ssh/known_hosts"; then
      ssh-keygen -R "desktop_ip_address" > /dev/null
    fi
    chezmoi update
  }
else
    alias chu="chezmoi update"
fi
```

Adding `StrictHostKeyChecking=accept-new` for the desktop entry in `~/.ssh/config` would also avoid an extra manual input in the terminal to accept the new key. I'm not sure about the security implications of this instead of `ask` which is the default, so I think 'm not going to use it.

**Comment by twpayne** [comment]
As @halostatue says, this is a git issue. It has nothing to do with chezmoi. Please ask on the git support forums.

**Comment by Lockszmith-GH** [comment]
As twpayne mentioned, not a chezmoi issue.
But take a look at [this](https://stackoverflow.com/a/74410573/799379), it might be what you're looking for.
